crypto/signatures.go

```diff
@@ -25,26 +25,74 @@ var (
 	ErrInvalidSignature = errors.New("invalid signature")
 )
 
-// GenerateKeyPair creates a public-secret keypair that can be used to sign and
-// verify messages.
-func GenerateSignatureKeys() (sk SecretKey, pk PublicKey, err error) {
-	var entropy [EntropySize]byte
-	_, err = rand.Read(entropy[:])
+// KeyPairGenerator allows the caller to generate public-secret key pairs.
+type KeyPairGenerator interface {
+	Generate() (SecretKey, PublicKey, error)
+	GenerateDeterministic(entropy [EntropySize]byte) (SecretKey, PublicKey)
+}
+
+// entropySource allows the caller to retrieve an array of bytes populated to
+// random values.
+type entropySource interface {
+	getEntropy() ([EntropySize]byte, error)
+}
+
+// keyDeriver allows the caller to generate a public-secret key pair based on
+// provided entropy.
+type keyDeriver interface {
+	deriveKeyPair([EntropySize]byte) (SecretKey, PublicKey)
+}
+
+// stdGenerator is an implementation of KeyPairGenerator, allowing the caller
+// to generate public-secret key pairs.
+type stdGenerator struct {
+	es entropySource
+	kd keyDeriver
+}
```

Comment:
you can sandwich this chain of types inside of the 

```
type(
)
```

declaration. I usually do that when there are lots of types.